packages/edit-post/src/components/editor-initialization/listener-hooks.js

```diff
@@ -0,0 +1,106 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, useDispatch } from '@wordpress/data';
+import { useEffect, useRef } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import {
+	STORE_KEY,
+	VIEW_AS_LINK_SELECTOR,
+	VIEW_AS_PREVIEW_LINK_SELECTOR,
+} from '../../store/constants';
+
+/**
+ * This listener hook monitors for block selection and triggers the appropriate
+ * sidebar state.
+ *
+ * @param {number} postId  The current post id.
+ */
+export const useBlockSelectionListener = ( postId ) => {
+	const {
+		hasBlockSelection,
+		isEditorSidebarOpened,
+	} = useSelect(
+		( select ) => ( {
+			hasBlockSelection: !! select(
+				'core/block-editor'
+			).getBlockSelectionStart(),
+			isEditorSidebarOpened: select( STORE_KEY ).isEditorSidebarOpened(),
+		} ),
+		[ postId ]
+	);
+
+	const { openGeneralSidebar } = useDispatch( STORE_KEY );
+
+	useEffect( () => {
+		if ( ! isEditorSidebarOpened ) {
+			return;
+		}
+		if ( hasBlockSelection ) {
+			openGeneralSidebar( 'edit-post/block' );
+		} else {
+			openGeneralSidebar( 'edit-post/document' );
+		}
+	}, [ hasBlockSelection, isEditorSidebarOpened ] );
+};
+
+/**
+ * This listener hook is used to monitor viewport size and adjust the sidebar
+ * accordingly.
+ *
+ * @param {number} postId  The current post id.
+ */
+export const useAdjustSidebarListener = ( postId ) => {
+	const { isSmall, sidebarToReOpenOnExpand } = useSelect(
+		( select ) => ( {
+			isSmall: select( 'core/viewport' ).isViewportMatch( '< medium' ),
+			sidebarToReOpenOnExpand: select( STORE_KEY ).getActiveGeneralSidebarName(),
+		} ),
+		[ postId ]
+	);
+
+	const { openGeneralSidebar, closeGeneralSidebar } = useDispatch( STORE_KEY );
+
+	const previousOpenedSidebar = useRef( '' );
```

Comment:
I might think `null` serves better as representing an unset value (this is also what was used previously). Unless for some reason `null` can't be used as the argument of `useRef`?

Reply:
I used empty string because of a habit I have of using the same type for empty values (`previousOpenedSidebar.current` would have a string value).  I think it's a good signal of what type to expect?

Reply:
> I used empty string because of a habit I have of using the same type for empty values (`previousOpenedSidebar.current` would have a string value). I think it's a good signal of what type to expect?

Hm, to me, we're representing an explicitly empty value, which is the purpose that a null type serves ([reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/null)), and isn't really a semantic guarantee of an empty string.

Reply:
The other option we may have here is to not set any value (just initialize with `useRef`)?

Reply:
Although that may mean `previousOpenedSidebar.current` will be undefined (I'd have to check).

Reply:
Yeah, I'd still think `null` would be preferred over `undefined`, as the distinction between the two is in intentionality / explicitness of emptiness (vs. merely "unset").

Reply:
However, in your reference is stated:
>  represents the intentional absence of any object value

And the accompanying example illustrates where null is used to reference an empty value for something that would return an object otherwise.  That's typically how I use `null` as the initial value for a variable, when the type returned would be an object as opposed to a primitive.

Reply:
> when the type returned would be an object as opposed to a primitive

I'm not sure I follow. You're suggesting to distinguish it based on whether it's `typeof 'object'` vs. some other primitive type?

Reply:
Right.  In this case `previousOpenedSidebar.current` will always return a string.  So it's empty representation would be an empty string (keeping type equality).  If `previousOpenedSidebar.current` is expected to return an object, then it's empty representation would be `null` (which seems to be the use-case for how MDN describes using `null`). 

On line 75 I also followed the same line of thought by resetting `previousOpenedSidebar.current` to an empty string:

https://github.com/WordPress/gutenberg/blob/521664cae2e2d637af4206bcc5d24d032b1f4e47/packages/edit-post/src/components/editor-initialization/listener-hooks.js#L75

I think `null` would be also be an appropriate initial value to use in the case where a value may represent any number of data types.